### PR TITLE
add freeze for development and full range for install

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Closes #12383
set all max versions of today's latest released version and define dependabot to open PR when a new release appears...
in setup add logic for removing this upper bound unless it is commented in as strict

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca @akihironitta @borda @tchaton @rohitgr7

### CODE REVIEW

File: .github/dependabot.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    labels:
+      - "dependencies"
+    pull-request-branch-name:
+      # Separate sections of the branch name with a hyphen
+      # for example, `dependabot-npm_and_yarn-next_js-acorn-6.4.1`
+      separator: "-"
+    # Allow up to 10 open pull requests for pip dependencies
+    open-pull-requests-limit: 10
+    reviewers:
+      - "octocat"
```
Comment (carmocca) [inline .github/dependabot.yml:23]:
```suggestion
```

File: .github/dependabot.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  # Enable version updates for python
+  - package-ecosystem: "pip"
+    # Look for a `requirements` in the `root` directory
+    directory: "/"
+    # Check for updates once a week
+    schedule:
+      interval: "weekly"
+    # Labels on pull requests for version updates only
+    labels:
+      - "dependencies"
```
Comment (carmocca) [inline .github/dependabot.yml:15]:
```suggestion
      - "ci"
```

File: .github/dependabot.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# Basic dependabot.yml file with
+# minimum configuration for two package managers
+
+version: 2
+updates:
+  # Enable version updates for python
+  - package-ecosystem: "pip"
+    # Look for a `requirements` in the `root` directory
+    directory: "/"
```
Comment (carmocca) [inline .github/dependabot.yml:9]: Won't this only modify the root requirements.txt? We need to tell it to modify all

File: requirements/strategies.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-fairscale>=0.4.5
+fairscale>=0.4.5, <=0.4.6
 deepspeed<0.6.0
 horovod>=0.21.2,!=0.24.0  # no need to install with [pytorch] as pytorch is already installed
 hivemind>=1.0.1; sys_platform == 'linux'
```
Comment (carmocca) [inline requirements/strategies.txt:4]: These have no max pin. On purpose?

Comment (Borda) [reply requirements/strategies.txt:4]: probably would need to come with better reg expression, so let's first evaluate the core of this PR and then we can finish this last one :)

**Review by awaelchli — APPROVED** [review]
awesome!

File: pytorch_lightning/setup_tools.py [hunk 5]
```diff
@@ -19,7 +19,9 @@
 _PROJECT_ROOT = os.path.dirname(os.path.dirname(__file__))
 
 
-def _load_requirements(path_dir: str, file_name: str = "requirements.txt", comment_char: str = "#") -> List[str]:
+def _load_requirements(
```
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/setup_tools.py:22]: when you install PL as a python package our requirements listed in files need to be loaded and passed to the setup function... till now we just dropped all comments and HTTP links, but now we also remove all upper bound version limitations unless it is intended (the commend after package includes word `strict`)

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/setup_tools.py:22]: Okay I see. So basically if the requirement says `x >= 1.5, <1.7`, we will install as if `x >= 1.5`. Sounds good! Do we actually have examples where the comment includes "strict"?

**Review by awaelchli — APPROVED** [review]

File: requirements/strategies.txt (hunk 4 above)
Comment (Borda) [reply requirements/strategies.txt]: fixed :rabbit:

File: pytorch_lightning/setup_tools.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/setup_tools.py:22]: > Okay I see. So basically if the requirement says `x >= 1.5, <1.7`, we will install as if x >= 1.5

the pattern for replacement is `<=` not `<` as the dependabot would list the last feasible version

> Do we actually have examples where the comment includes "strict"?

not yet as we do not have such case yet :/

File: .github/dependabot.yml (hunk 3 above)
Comment (Borda) [reply .github/dependabot.yml]: well, seems there is only one folder, so moving `./requirements.txt` to `./requirements/base.txt` and keeping the original `./requirements.txt` ad link

File: requirements/base.txt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+numpy>=1.17.2, <=1.22.3
+torch>=1.8.*, <=1.11.0
+tqdm>=4.57.0, <=4.63.0
+PyYAML>=5.4, <=6.0
+fsspec[http]>=2021.05.0, !=2021.06.0, <=2022.2.0
+tensorboard>=2.2.0, <=2.8.0
+torchmetrics>=0.4.1, <=0.7.2
+pyDeprecate>=0.3.1, <0.4.0, <=0.3.2
```
Comment (akihironitta) [inline requirements/base.txt:8]: This is now redundant? https://github.com/Borda/pyDeprecate/releases

```suggestion
pyDeprecate>=0.3.1, <=0.3.2
```

File: requirements/extra.txt [hunk 7]
```diff
@@ -1,9 +1,9 @@
...
-matplotlib>3.1
-torchtext>=0.9.*
-omegaconf>=2.0.5
-hydra-core>=1.0.5
-jsonargparse[signatures]>=4.7.1
-gcsfs>=2021.5.0
-rich>=10.2.2,!=10.15.*
+matplotlib>3.1, <=3.5.1
+torchtext>=0.9.*, <=0.12.0
+omegaconf>=2.0.5, <=2.1.*
```
Comment (akihironitta) [inline requirements/extra.txt:5]: noob question: How is it different from specifying the exact version? e.g. `<=2.1.*` vs. `<=2.1.2`

Not sure why some use `.*` and others use exact versions.

**Review by akihironitta — COMMENTED** [review]
The change in `setup_tools.py` is very interesting 🚀

File: requirements/extra.txt (hunk 7 above)
Comment (Borda) [reply requirements/extra.txt:5]: `.*` includes all bugfixes...
`<=2.1.*` is the same as `<2.2`

Comment (akihironitta) [reply requirements/extra.txt:5]: Should all upper bounds be written that way then?

For example,
```diff
- matplotlib>3.1, <=3.5.1
- torchtext>=0.9.*, <=0.12.0
+ matplotlib>3.1, <=3.5.*
+ torchtext>=0.9.*, <=0.12.*
```

Comment (akihironitta) [reply requirements/extra.txt:5]: I think it's nothing critical, but I'm just curious why some are written so while others aren't.

Comment (Borda) [reply requirements/extra.txt:5]: we can convert them to a specific version :)

**Review by carmocca — APPROVED** [review]

File: .github/dependabot.yml (hunk 3 above)
Comment (akihironitta) [reply .github/dependabot.yml]: SGTM!

File: pytorch_lightning/setup_tools.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (akihironitta) [reply pytorch_lightning/setup_tools.py:22]: This is nice! Please correct me if I'm wrong, but with this logic, we don't need to change anything in the requirements files (i.e., we don't need to unpin the upper-bounded versions) when releasing, right?

File: requirements/extra.txt (hunk 7 above)
Comment (akihironitta) [reply requirements/extra.txt:5]: IMHO, we should have specific versions for all packages to further reduce the possibility of something getting broken.

File: pytorch_lightning/setup_tools.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/setup_tools.py:22]: @kaushikb11 ^^

File: pytorch_lightning/setup_tools.py [hunk 8]
```diff
@@ -31,12 +33,17 @@ def _load_requirements(path_dir: str, file_name: str = "requirements.txt", comme
...
         # skip directly installed dependencies
-        if ln.startswith("http") or "@http" in ln:
+        if not req or req.startswith("http") or "@http" in req:
             continue
-        if ln:  # if requirement is not empty
-            reqs.append(ln)
+        # remove version restrictions unless they are strict
+        if unfreeze and "<" in req and "strict" not in comment:
+            req = re.sub(r",? *<= *[\d\.\*]+", "", req).strip()
+            req = req[:-1] if req[-1] == "," else req
```
Comment (Borda) [inline pytorch_lightning/setup_tools.py:45]:
```suggestion
```

File: pytorch_lightning/setup_tools.py [hunk 9]
```diff
@@ -31,12 +33,16 @@ def _load_requirements(path_dir: str, file_name: str = "requirements.txt", comme
...
+        comment = ln[ln.index(comment_char) :] if comment_char in ln else ""
+        req = ln.strip()
         # skip directly installed dependencies
-        if ln.startswith("http") or "@http" in ln:
+        if not req or req.startswith("http") or "@http" in req:
             continue
-        if ln:  # if requirement is not empty
-            reqs.append(ln)
+        # remove version restrictions unless they are strict
+        if unfreeze and "<" in req and "strict" not in comment:
```
Comment (Borda) [inline pytorch_lightning/setup_tools.py:43]:
```suggestion
        if unfreeze and "<=" in req and "strict" not in comment:
```

**Review by Borda — COMMENTED** [review]
->

File: requirements/strategies.txt [hunk 10]
```diff
@@ -1,4 +1,5 @@
-fairscale>=0.4.5
+fairscale>=0.4.5, <=0.4.6
 deepspeed<0.6.0
-horovod>=0.21.2,!=0.24.0  # no need to install with [pytorch] as pytorch is already installed
-hivemind>=1.0.1; sys_platform == 'linux'
+# no need to install with [pytorch] as pytorch is already installed
+horovod>=0.21.2,!=0.24.0, <=0.24.3
```
[the remark this answers is not in the document]
Comment (Borda) [reply requirements/strategies.txt:4]: it was too long line, but yes, we can move it back...

**Pull request merged**